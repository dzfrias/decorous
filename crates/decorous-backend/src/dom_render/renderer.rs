use decorous_frontend::{
    ast::{Attribute, AttributeValue, CollapsedChildrenType, Node, NodeType},
    utils, DeclaredVariables, FragmentMetadata,
};
use std::{borrow::Cow, io};

use crate::codegen_utils;

pub trait Renderer<T>
where
    T: io::Write,
{
    fn init(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()>;
    fn create(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()>;
    fn mount(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()>;
    fn update(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()>;
    fn detach(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()>;
}

impl<'a, T> Renderer<T> for Node<'a, FragmentMetadata>
where
    T: io::Write,
{
    fn init(&self, f: &mut T, _toplevel_vars: &DeclaredVariables) -> io::Result<()> {
        match self.node_type() {
            NodeType::Element(element) => {
                writeln!(f, "let e{};", self.metadata().id())?;

                if element.inner_collapsed().is_some() {
                    return Ok(());
                }
                for child in element.children() {
                    child.init(f, _toplevel_vars)?;
                }
                Ok(())
            }
            NodeType::Text(_) | NodeType::Mustache(_) => {
                writeln!(f, "let e{};", self.metadata().id())
            }
            NodeType::Comment(_) => Ok(()),
            NodeType::SpecialBlock(_) => todo!(),
        }
    }

    fn create(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()> {
        match self.node_type() {
            NodeType::Element(element) => {
                writeln!(
                    f,
                    "e{} = document.createElement(\"{}\");",
                    self.metadata().id(),
                    element.tag()
                )?;

                for attr in element.attrs() {
                    match attr {
                        Attribute::KeyValue(key, value) => match value {
                            Some(AttributeValue::Literal(literal)) => {
                                let (s, use_quotes) = collapse_whitespace(literal);
                                writeln!(
                                    f,
                                    "e{}.setAttribute(\"{key}\", {}{}{1});",
                                    use_quotes.then_some("\"").unwrap_or_default(),
                                    s,
                                    self.metadata().id(),
                                )?;
                            }
                            Some(AttributeValue::JavaScript(js)) => writeln!(
                                f,
                                "e{}.setAttribute(\"{key}\", {});",
                                self.metadata().id(),
                                codegen_utils::replace_namerefs(
                                    js,
                                    &utils::get_unbound_refs(js),
                                    toplevel_vars
                                )
                            )?,
                            None => writeln!(
                                f,
                                "e{}.setAttribute(\"{key}\", \"\");",
                                self.metadata().id(),
                            )?,
                        },
                        Attribute::EventHandler(event_handler) => {
                            writeln!(
                                f,
                                "e{}.addEventListener(\"{}\", {});",
                                self.metadata().id(),
                                event_handler.event(),
                                codegen_utils::replace_namerefs(
                                    event_handler.expr(),
                                    &utils::get_unbound_refs(event_handler.expr()),
                                    toplevel_vars
                                )
                            )?;
                        }
                        Attribute::Binding(_binding) => todo!(),
                    }
                }

                if let Some(collapsed) = element.inner_collapsed() {
                    return match collapsed {
                        CollapsedChildrenType::Text(t) => {
                            let (s, use_qutoes) = collapse_whitespace(t);
                            writeln!(
                                f,
                                "e{}.textContent = {}{s}{1};",
                                self.metadata().id(),
                                use_qutoes.then_some("\"").unwrap_or_default()
                            )
                        }
                        CollapsedChildrenType::Html(html) => {
                            writeln!(f, "e{}.innerHtml = \"{html}\"", self.metadata().id())
                        }
                    };
                }
                for child in element.children() {
                    child.create(f, toplevel_vars)?;
                }

                Ok(())
            }
            NodeType::Text(text) => {
                let (s, use_qutoes) = collapse_whitespace(text);
                writeln!(
                    f,
                    "e{} = document.createTextNode({}{s}{1});",
                    self.metadata().id(),
                    use_qutoes.then_some("\"").unwrap_or_default()
                )
            }
            NodeType::Mustache(mustache) => {
                let new_text = codegen_utils::replace_namerefs(
                    mustache,
                    &utils::get_unbound_refs(mustache),
                    toplevel_vars,
                );
                writeln!(
                    f,
                    "e{} = document.createTextNode({new_text});",
                    self.metadata().id()
                )
            }
            NodeType::Comment(_) => Ok(()),
            NodeType::SpecialBlock(_) => todo!(),
        }
    }

    fn mount(&self, f: &mut T, _toplevel_vars: &DeclaredVariables) -> io::Result<()> {
        if matches!(self.node_type(), NodeType::Comment(_)) {
            return Ok(());
        }

        if let Some(parent_id) = self.metadata().parent_id() {
            writeln!(f, "e{parent_id}.appendChild(e{});", self.metadata().id())?;
        } else {
            writeln!(f, "target.appendChild(e{});", self.metadata().id())?;
        }

        match self.node_type() {
            NodeType::Element(element) if element.inner_collapsed().is_none() => {
                for child in element.children() {
                    child.mount(f, _toplevel_vars)?;
                }
            }
            NodeType::SpecialBlock(_) => todo!(),
            _ => {}
        }

        Ok(())
    }

    fn update(&self, f: &mut T, toplevel_vars: &DeclaredVariables) -> io::Result<()> {
        match self.node_type() {
            NodeType::Element(elem) => {
                for attr in elem.attrs() {
                    // TODO: Update to use dirty
                    if let Attribute::KeyValue(key, Some(AttributeValue::JavaScript(js))) = attr {
                        let unbound = utils::get_unbound_refs(js);
                        let dirty_indices = codegen_utils::calc_dirty(&unbound, toplevel_vars);
                        let replacement =
                            codegen_utils::replace_namerefs(js, &unbound, toplevel_vars);
                        writeln!(
                            f,
                            "if ({dirty_indices}) e{}.setAttribute(\"{key}\", {replacement});",
                            self.metadata().id(),
                        )?;
                    }
                }

                if elem.inner_collapsed().is_none() {
                    for child in elem.children() {
                        child.update(f, toplevel_vars)?;
                    }
                }
                Ok(())
            }
            NodeType::Mustache(mustache) => {
                let unbound = utils::get_unbound_refs(mustache);
                let dirty_indices = codegen_utils::calc_dirty(&unbound, toplevel_vars);
                let new_text = codegen_utils::replace_namerefs(mustache, &unbound, toplevel_vars);
                writeln!(
                    f,
                    "if ({dirty_indices}) e{}.data = {new_text};",
                    self.metadata().id()
                )
            }
            _ => Ok(()),
        }
    }

    fn detach(&self, f: &mut T, _toplevel_vars: &DeclaredVariables) -> io::Result<()> {
        if matches!(self.node_type(), NodeType::Comment(_)) {
            return Ok(());
        }

        // Filter out non-root components
        if self.metadata().parent_id().is_some() {
            return Ok(());
        }

        writeln!(f, "e{}.parentNode.removeChild(e{0});", self.metadata().id())
    }
}

// Boolean also returned because I really didn't want to allocate a new String here... maybe
// there's a better way? The bool denotes if the caller should use quotes when surrounding the
// text.
fn collapse_whitespace(input: &str) -> (Cow<str>, bool) {
    if input == "\n" {
        (Cow::Borrowed(" "), true)
    } else if input.contains('\n') {
        (Cow::Owned(format!("String.raw`{input}`")), false)
    } else {
        (Cow::Borrowed(input), true)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use decorous_frontend::ast::{Element, Location, Mustache, Text};
    use rslint_parser::{parse_text, SmolStr};

    macro_rules! test_lifecycle {
        ($node:expr, $cycle_func:ident, $expected:expr, $unbound_refs:expr) => {
            let mut out = Vec::new();
            assert!($node.$cycle_func(&mut out, $unbound_refs).is_ok());
            assert_eq!($expected, String::from_utf8(out).unwrap());
        };
    }

    #[test]
    fn can_write_basic_text_nodes() {
        let node = Node::new(
            NodeType::Text(Text("hello")),
            FragmentMetadata::new(0, None, Location::default()),
        );

        test_lifecycle!(node, init, "let e0;\n", &DeclaredVariables::new());
        test_lifecycle!(
            node,
            create,
            "e0 = document.createTextNode(\"hello\");\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            mount,
            "target.appendChild(e0);\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            detach,
            "e0.parentNode.removeChild(e0);\n",
            &DeclaredVariables::new()
        );
    }

    #[test]
    fn basic_elements_and_html_are_collapsed() {
        let node = Node::new(
            NodeType::Element(Element::new(
                "div",
                vec![],
                vec![
                    Node::new(
                        NodeType::Text(Text("text")),
                        FragmentMetadata::new(1, Some(0), Location::default()),
                    ),
                    Node::new(
                        NodeType::Element(Element::new("div", vec![], vec![])),
                        FragmentMetadata::new(2, Some(0), Location::default()),
                    ),
                ],
            )),
            FragmentMetadata::new(0, None, Location::default()),
        );

        test_lifecycle!(node, init, "let e0;\n", &DeclaredVariables::new());
        test_lifecycle!(
            node,
            create,
            "e0 = document.createElement(\"div\");\ne0.innerHtml = \"text<div></div>\"\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            mount,
            "target.appendChild(e0);\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            detach,
            "e0.parentNode.removeChild(e0);\n",
            &DeclaredVariables::new()
        );
    }

    #[test]
    fn can_write_mustache_tags() {
        let js = parse_text("(hi, hi)", 0).syntax().first_child().unwrap();
        let node = Node::new(
            NodeType::Mustache(Mustache(js)),
            FragmentMetadata::new(0, None, Location::default()),
        );

        let declared = {
            let mut vars = DeclaredVariables::new();
            vars.insert_var(SmolStr::new("hi"));
            vars
        };
        test_lifecycle!(node, init, "let e0;\n", &declared);
        test_lifecycle!(
            node,
            create,
            "e0 = document.createTextNode((ctx[0], ctx[0]));\n",
            &declared
        );
        test_lifecycle!(node, mount, "target.appendChild(e0);\n", &declared);
        test_lifecycle!(
            node,
            update,
            "if (dirty[0] & 1) e0.data = (ctx[0], ctx[0]);\n",
            &declared
        );
        test_lifecycle!(node, detach, "e0.parentNode.removeChild(e0);\n", &declared);
    }

    #[test]
    fn text_with_only_one_text_node_as_child_is_collapsed() {
        let node = Node::new(
            NodeType::Element(Element::new(
                "span",
                vec![],
                vec![Node::new(
                    NodeType::Text(Text("hello")),
                    FragmentMetadata::new(1, Some(0), Location::default()),
                )],
            )),
            FragmentMetadata::new(0, None, Location::default()),
        );

        test_lifecycle!(node, init, "let e0;\n", &DeclaredVariables::new());
        test_lifecycle!(
            node,
            create,
            "e0 = document.createElement(\"span\");\ne0.textContent = \"hello\";\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            mount,
            "target.appendChild(e0);\n",
            &DeclaredVariables::new()
        );
        test_lifecycle!(
            node,
            detach,
            "e0.parentNode.removeChild(e0);\n",
            &DeclaredVariables::new()
        );
    }

    #[test]
    fn dirty_items_are_in_conditional() {
        let js = parse_text("(hello, test)", 0)
            .syntax()
            .first_child()
            .unwrap();
        let node = Node::new(
            NodeType::Mustache(Mustache(js)),
            FragmentMetadata::new(0, None, Location::default()),
        );

        let declared = {
            let mut vars = DeclaredVariables::new();
            vars.insert_var(SmolStr::new("hello"));
            vars.insert_var(SmolStr::new("test"));
            vars
        };
        test_lifecycle!(
            node,
            update,
            "if (dirty[0] & 3) e0.data = (ctx[0], ctx[1]);\n",
            &declared
        );
    }
}
