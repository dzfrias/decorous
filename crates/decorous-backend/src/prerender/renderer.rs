use decorous_frontend::ast::{Attribute, AttributeValue};
use rslint_parser::{SmolStr, SyntaxNode};
use std::collections::HashMap;
use std::fmt::Write;
use std::io;

use decorous_frontend::{
    ast::{Node, NodeType},
    utils, Component, FragmentMetadata,
};

use crate::{codegen_utils, replace};

#[derive(Debug)]
struct TransientNodeData<'a> {
    pub mustache_data: Vec<(usize, SyntaxNode)>,
    pub attr_data: Vec<AttrData<'a>>,
    pub event_listeners: Vec<Listener<'a>>,
}

#[derive(Debug)]
struct AttrData<'a> {
    pub attr: &'a str,
    pub js: SyntaxNode,
    pub elems_idx: usize,
}

#[derive(Debug)]
struct Listener<'a> {
    pub event: &'a str,
    pub js: SyntaxNode,
    pub elems_idx: usize,
}

#[derive(Debug)]
enum ElementData {
    Element { id: u32, child_pos: usize },
    Mustache { id: u32 },
    Attribute { id: u32 },
    Listener { id: u32 },
}

impl ElementData {
    pub fn id(&self) -> u32 {
        match self {
            Self::Element { id, .. } => *id,
            Self::Mustache { id } => *id,
            Self::Attribute { id } => *id,
            Self::Listener { id } => *id,
        }
    }
}

#[derive(Debug)]
pub struct Renderer<'a> {
    component: &'a Component<'a>,
    ctx_body: String,
    update_body: String,
    html: String,
    // Contains information for the `elems` array in the output JS. The first element of the tuple
    // corresponds to the id of the element. The second (if set) contains the index of the child
    // node to get a reference to. This corresponds to the index of a mustache tag in an element.
    elems: Vec<ElementData>,
    id_overwrites: HashMap<u32, SmolStr>,
}

impl<'a> Renderer<'a> {
    pub fn new(component: &'a Component<'a>) -> Self {
        Self {
            component,
            ctx_body: String::new(),
            update_body: String::new(),
            html: String::new(),
            elems: vec![],
            id_overwrites: HashMap::new(),
        }
    }

    pub fn render<T, U>(mut self, js_out: &mut T, html_out: &mut U) -> io::Result<()>
    where
        T: io::Write,
        U: io::Write,
    {
        self.build_render_sections();
        let elems = self.build_elems();

        // Write finished segments
        write!(
            js_out,
            include_str!("./template.js"),
            dirty_items = ((self.component.declared_vars().all_vars().len() + 7) / 8),
            elems = elems,
            ctx_body = self.ctx_body,
            update_body = self.update_body
        )?;
        write!(html_out, "{}", self.html)?;

        Ok(())
    }

    fn build_render_sections(&mut self) {
        self.start_ctx_body();
        for node in self.component.descendents() {
            if let Some(data) = self.get_node_data(node) {
                self.build_ctx_body(&data);
                self.build_update_body(&data);
            }
        }
        self.end_ctx_body();
        writeln!(self.html, "<script defer src=\"./out.js\"></script>")
            .expect("string format should not fail");
        self.build_html();
    }

    fn build_elems(&self) -> String {
        let mut elems = String::new();
        for element_data in &self.elems {
            let string_id = SmolStr::new(element_data.id().to_string());
            let id = self
                .id_overwrites
                .get(&element_data.id())
                .unwrap_or(&string_id);
            match element_data {
                ElementData::Element { child_pos, .. } => {
                    write!(
                        elems,
                        "replace(document.getElementById(\"{id}\").childNodes[{child_pos}]),"
                    )
                    .expect("string format should not fail");
                }
                ElementData::Mustache { .. } => {
                    write!(elems, "replace(document.getElementById(\"{id}\"))")
                        .expect("string format should not fail");
                }
                ElementData::Attribute { .. } | ElementData::Listener { .. } => {
                    write!(elems, "document.getElementById(\"{id}\"),")
                        .expect("string format should not fail");
                }
            }
        }
        elems
    }

    // One of the core methods of the renderer. It receives a node. If the node is not an element,
    // it returns None. Otherwise, it grabs some data about the node. This data contains
    // instructions for how to update the element. If it has not special mustache tags, this data
    // will contain nothing. With dynamic attributes, this method returns information about that
    // attribute, such as it's name, js-generated value, and index in the `elems` array (on the
    // Javascript side). With a dynamic text node, it contains that node's js-generated value and
    // it's index in the `elems` array.
    //
    // It also mutates `self.elems`.
    fn get_node_data(
        &mut self,
        node: &'a Node<'_, FragmentMetadata>,
    ) -> Option<TransientNodeData<'a>> {
        match node.node_type() {
            NodeType::Element(elem) => {
                let mut attr_data = vec![];
                let mut listeners = vec![];
                for attr in elem.attrs() {
                    match attr {
                        Attribute::KeyValue(key, Some(AttributeValue::JavaScript(js))) => {
                            self.elems.push(ElementData::Attribute {
                                id: node.metadata().id(),
                            });
                            let idx = self.elems.len() - 1;
                            attr_data.push(AttrData {
                                attr: *key,
                                js: js.clone(),
                                elems_idx: idx,
                            });
                        }
                        Attribute::EventHandler(event_handler) => {
                            self.elems.push(ElementData::Listener {
                                id: node.metadata().id(),
                            });
                            let idx = self.elems.len() - 1;
                            listeners.push(Listener {
                                event: event_handler.event(),
                                js: event_handler.expr().clone(),
                                elems_idx: idx,
                            });
                        }
                        // TODO: Possibly bindings?
                        _ => {}
                    }
                }

                let mut elems_indices = vec![];
                for (pos, mustache) in
                    elem.children()
                        .iter()
                        .enumerate()
                        .filter_map(|(idx, node)| match node.node_type() {
                            NodeType::Mustache(mustache) => Some((idx, mustache)),
                            NodeType::Text(_) | NodeType::Comment(_) | NodeType::Element(_) => None,
                            NodeType::SpecialBlock(_) => todo!("special blocks"),
                            NodeType::Error => panic!("should not have error nodes at this point"),
                        })
                {
                    self.elems.push(ElementData::Element {
                        id: node.metadata().id(),
                        child_pos: pos,
                    });
                    let idx = self.elems.len() - 1;
                    elems_indices.push((idx, mustache.clone()));
                }
                Some(TransientNodeData {
                    mustache_data: elems_indices,
                    attr_data,
                    event_listeners: listeners,
                })
            }
            NodeType::Mustache(mustache) if node.metadata().parent_id().is_none() => {
                self.elems.push(ElementData::Mustache {
                    id: node.metadata().id(),
                });
                Some(TransientNodeData {
                    mustache_data: vec![(self.elems.len() - 1, mustache.clone())],
                    attr_data: vec![],
                    event_listeners: vec![],
                })
            }
            _ => None,
        }
    }

    fn build_ctx_body(&mut self, data: &TransientNodeData<'_>) {
        for (idx, js) in &data.mustache_data {
            writeln!(self.ctx_body, "elems[{idx}].data = {js};")
                .expect("writing to string format should not fail");
        }

        for AttrData {
            attr,
            js,
            elems_idx,
        } in &data.attr_data
        {
            writeln!(
                self.ctx_body,
                "elems[{elems_idx}].setAttribute(\"{attr}\", {js});"
            )
            .expect("writing to format string should not fail");
        }

        for Listener {
            event,
            js,
            elems_idx,
        } in &data.event_listeners
        {
            let replaced = replace::replace_assignments(
                js,
                &utils::get_unbound_refs(js),
                self.component.declared_vars(),
            );
            writeln!(
                self.ctx_body,
                "elems[{elems_idx}].addEventListener(\"{event}\", {replaced});"
            )
            .expect("writing to format string should not fail");
        }
    }

    fn build_update_body(&mut self, data: &TransientNodeData<'_>) {
        for (idx, js) in &data.mustache_data {
            let unbound = utils::get_unbound_refs(js);
            let dirty_indices = codegen_utils::calc_dirty(&unbound, self.component.declared_vars());
            let replaced = replace::replace_namerefs(&js, &unbound, self.component.declared_vars());
            writeln!(
                self.update_body,
                "if ({dirty_indices}) elems[{idx}].data = {replaced};",
            )
            .expect("writing to string format should not fail");
        }

        for AttrData {
            attr,
            js,
            elems_idx,
        } in &data.attr_data
        {
            let unbound = utils::get_unbound_refs(js);
            let dirty_indices = codegen_utils::calc_dirty(&unbound, self.component.declared_vars());
            let replaced = replace::replace_namerefs(&js, &unbound, self.component.declared_vars());
            writeln!(
                self.update_body,
                "if ({dirty_indices}) elems[{elems_idx}].setAttribute(\"{attr}\", {replaced});",
            )
            .expect("writing to format string should not fail");
        }
    }

    fn start_ctx_body(&mut self) {
        for node in self.component.toplevel_nodes() {
            if node.substitute_assign_refs {
                let replacement = replace::replace_assignments(
                    &node.node,
                    &utils::get_unbound_refs(&node.node),
                    self.component.declared_vars(),
                );
                writeln!(self.ctx_body, "{}", replacement)
                    .expect("string formatting should not fail");
            } else {
                writeln!(self.ctx_body, "{}", node.node)
                    .expect("string formatting should not fail");
            }
        }
    }

    fn end_ctx_body(&mut self) {
        let mut ctx = vec![""; self.component.declared_vars().all_vars().len()];
        for (name, idx) in self.component.declared_vars().all_vars() {
            ctx[*idx as usize] = name;
        }
        writeln!(self.ctx_body, "return [{}];", ctx.join(","))
            .expect("string format should not fail");
    }

    // Builds the HTML part of the output
    fn build_html(&mut self) {
        for node in self.component.fragment_tree() {
            self.build_html_for_node(node);
        }
    }

    fn build_html_for_node(&mut self, node: &Node<'_, FragmentMetadata>) {
        match node.node_type() {
            NodeType::Text(t) => write!(self.html, "{t}").expect("string format should not fail"),
            NodeType::Comment(t) => {
                write!(self.html, "<!--{t}-->").expect("string format should not fail")
            }
            NodeType::Mustache(_) if node.metadata().parent_id().is_none() => {
                write!(self.html, "<span id=\"{}\"></span>", node.metadata().id())
                    .expect("string format should not fail")
            }
            // Write a span. This segments the childNodes of the element so that an actual text
            // node can be placed here.
            NodeType::Mustache(_) => {
                write!(self.html, "<span></span>").expect("string format should not fail")
            }
            NodeType::Element(elem) => {
                write!(self.html, "<{}", elem.tag()).expect("string format should not fail");
                let mut has_dynamic = false;
                for attr in elem.attrs() {
                    match attr {
                        Attribute::KeyValue(key, Some(AttributeValue::Literal(literal))) => {
                            if *key == "id" {
                                // The user has set this id, so use it from now on instead of a
                                // pre-generated one
                                self.id_overwrites
                                    .insert(node.metadata().id(), literal.into());
                            }
                            write!(self.html, " {key}=\"{literal}\"")
                                .expect("string format should not fail");
                        }
                        // Do nothing. Dynamic attributes can't be baked statically into the HTML
                        Attribute::KeyValue(_, Some(AttributeValue::JavaScript(_)))
                        | Attribute::EventHandler(_) => has_dynamic = true,
                        Attribute::KeyValue(key, None) => {
                            write!(self.html, " {key}=\"\" ")
                                .expect("string format should not fail");
                        }
                        Attribute::Binding(_) => todo!(),
                    }
                }
                if (elem.has_immediate_mustache() || has_dynamic)
                    && !self.id_overwrites.contains_key(&node.metadata().id())
                {
                    write!(self.html, " id=\"{}\"", node.metadata().id())
                        .expect("string format should not fail");
                }
                write!(self.html, ">").expect("string format should not fail");
                for child in elem.children() {
                    self.build_html_for_node(child);
                }
                write!(self.html, "</{}>", elem.tag()).expect("string format should not fail");
            }
            NodeType::SpecialBlock(_) => todo!("special blocks"),
            NodeType::Error => panic!("should not try to render with error nodes"),
        }
    }
}

#[cfg(test)]
mod tests {
    use decorous_frontend::parse;

    use super::*;

    fn make_component(input: &str) -> Component {
        Component::new(parse(input).expect("should be valid input"))
    }

    macro_rules! test_render {
        ($($input:expr),+) => {
            $(
                let component = make_component($input);
                let mut js_out = Vec::new();
                let mut html_out = Vec::new();
                let renderer = Renderer::new(&component);
                renderer.render(&mut js_out, &mut html_out).unwrap();
                insta::assert_snapshot!(format!("{}\n---\n{}", String::from_utf8(js_out).unwrap(), String::from_utf8(html_out).unwrap()));
             )+
        };
    }

    #[test]
    fn can_write_basic_html_from_fragment_tree_ignoring_mustache_tags() {
        test_render!("#p Hello /p", "#div #p Hi /p Hello, {name} /div");
    }

    #[test]
    fn can_write_basic_js() {
        test_render!(
            "---js let x = 3; --- #p Hello, {x}! /p #button[@click={() => x = 444}] Click Me /button",
            "---js let x = 3; --- #p Hello, {x}! /p #button[@click={() => x = 444}] Click Me /button #p {x} /p"
        );
    }

    #[test]
    fn custom_ids_do_not_get_overriden() {
        test_render!("---js let x = 3;--- #p[id=\"custom\"] Hello, {x}! /p");
    }

    #[test]
    fn can_create_dynamic_attributes() {
        test_render!(
            "---js let x = 3; --- #p[class={x + 3}] Text /p",
            "---js let x = 3; --- #p[class={x + 3}] Hello {x} /p"
        );
    }

    #[test]
    fn supports_toplevel_mustache_tags() {
        test_render!("---js let x = 3; --- {x}");
    }

    #[test]
    fn multiple_variables_are_properly_in_dirty_buffer() {
        test_render!("---js let x = 0; let y = 0; --- #p {x} and {y} and {x + y} /p");
    }
}
