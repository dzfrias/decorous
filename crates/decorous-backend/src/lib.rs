use crate::renderer::Renderer;
use decorous_frontend::{utils, Component};
use itertools::Itertools;
use std::io;

mod renderer;
pub mod replace;

pub fn render<T: io::Write>(component: &Component, render_to: &mut T) -> io::Result<()> {
    macro_rules! render {
        ($lifecycle:ident) => {
            for node in component.fragment_tree() {
                node.$lifecycle(render_to, component.declared_vars())?;
            }
        };
    }

    // Hoisted syntax nodes should come first
    for hoist in component.hoist() {
        writeln!(render_to, "{hoist}")?;
    }

    writeln!(
        render_to,
        "const dirty = new Uint8Array(new ArrayBuffer({}));",
        // Ceiling division to get the amount of bytes needed in the ArrayBuffer
        ((component.declared_vars().len() + 8 - 1) / 8)
    )?;
    writeln!(render_to, "function make_fragment(ctx) {{")?;
    render!(init);

    writeln!(render_to, "return {{")?;
    writeln!(render_to, "c() {{")?;
    render!(create);
    writeln!(render_to, "}},")?;

    writeln!(render_to, "m(target) {{")?;
    render!(mount);
    writeln!(render_to, "}},")?;

    writeln!(render_to, "u(ctx, dirty) {{")?;
    render!(update);
    writeln!(render_to, "}},")?;

    writeln!(render_to, "d() {{")?;
    render!(detach);
    writeln!(render_to, "}},")?;

    writeln!(render_to, "}}")?;
    writeln!(render_to, "}}")?;

    writeln!(render_to, "function __init_ctx() {{")?;
    writeln!(
        render_to,
        "{}",
        component
            .toplevel_nodes()
            .iter()
            .map(|node| {
                if node.substitute_assign_refs {
                    replace::replace_assignments(
                        &node.node,
                        &utils::get_unbound_refs(&node.node),
                        component.declared_vars(),
                    )
                } else {
                    node.node.to_string()
                }
            })
            .join("\n")
    )?;

    let mut ctx = vec![""; component.declared_vars().len()];
    for (name, idx) in component.declared_vars() {
        ctx[*idx as usize] = name;
    }
    writeln!(render_to, "return [{}];", ctx.join(","))?;
    writeln!(render_to, "}}")?;

    writeln!(render_to, "const ctx = __init_ctx();")?;
    writeln!(render_to, "const fragment = make_fragment(ctx);")?;
    writeln!(render_to, "fragment.c();")?;
    writeln!(render_to, "fragment.m(document.getElementById(\"app\"));")?;
    writeln!(render_to, "let updating = false;")?;
    writeln!(
        render_to,
        "function __schedule_update(ctx_idx, val) {{
ctx[ctx_idx] = val;
dirty[Math.max(Math.ceil(ctx_idx / 8) - 1, 0)] |= 1 << (ctx_idx % 8);
if (updating) return;
updating = true;
Promise.resolve().then(() => {{
fragment.u(ctx, dirty);
updating = false;
dirty.fill(0);
}});
}}"
    )?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use decorous_frontend::{ParseError, Parser};

    use super::*;

    fn make_component(input: &str) -> Component {
        let parser = Parser::new(input);
        let parse_res = parser.parse();
        assert_eq!(Vec::<ParseError>::new(), parse_res.1);
        Component::new(parse_res.0)
    }

    macro_rules! test_render {
        ($input:expr) => {
            let component = make_component($input);
            let mut out = vec![];
            render(&component, &mut out).unwrap();

            println!("{}", String::from_utf8(out.clone()).unwrap());
            insta::assert_snapshot!(String::from_utf8(out).unwrap());
        };
    }

    #[test]
    fn basic_render_works() {
        test_render!(
            "<script>let x = 3; function remake_x() { x = 44; }</script><p>{`${x}hello`}</p>"
        );
    }

    #[test]
    fn render_with_attrs_works() {
        test_render!(
            "<script>let x = 3; function remake_x() { x = 44; }</script><div class={x}></div>"
        );
    }

    #[test]
    fn render_with_event_listeners_works() {
        test_render!(
            "<script>let x = 3;</script><p>{x}</p><button on:click={() => {if (x == 3) { x = 44; }} }>Clickme</button>"
        );
    }

    #[test]
    fn imports_are_hoisted_out_of_context_init() {
        test_render!("<script>import data from \"data\"; let x = 3;</script><p>{x}</p>");
    }

    #[test]
    fn everything_in_script_tag_is_in_context_init() {
        test_render!("<script>let x = 3; x = 4;</script>");
    }
}
