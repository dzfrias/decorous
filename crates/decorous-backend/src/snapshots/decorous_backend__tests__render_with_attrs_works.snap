---
source: crates/decorous-backend/src/lib.rs
expression: "String::from_utf8(out).unwrap()"
---
const dirty = new Uint8Array(new ArrayBuffer(1));
function make_fragment(ctx) {
let e0;
return {
c() {
e0 = document.createElement("div");
e0.setAttribute("class", ctx[0]);
},
m(target) {
target.appendChild(e0);
},
u(ctx, dirty) {
e0.setAttribute("class", ctx[0]);
},
d() {
e0.parentNode.removeChild(e0);
},
}
}
function __init_ctx() {
let x = 3;
function remake_x() { __schedule_update(0, x = 44); }
return [x,remake_x];
}
const ctx = __init_ctx();
const fragment = make_fragment(ctx);
fragment.c();
fragment.m(document.getElementById("app"));
let updating = false;
function __schedule_update(ctx_idx, val) {
ctx[ctx_idx] = val;
dirty[Math.max(Math.ceil(ctx_idx / 8) - 1, 0)] |= 1 << (ctx_idx % 8);
if (updating) return;
updating = true;
Promise.resolve().then(() => {
fragment.u(ctx, dirty);
updating = false;
dirty.fill(0);
});
}

